Detectors/TPC/qc/src/Clusters.cxx

```diff
@@ -0,0 +1,95 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "TH1.h"
+#include "TH2.h"
+#include "TFile.h"
+
+#include "TPCQC/Clusters.h"
+#include "TPCBase/Painter.h"
+#include "TPCBase/ROC.h"
+#include "TPCBase/Mapper.h"
+
+ClassImp(o2::tpc::qc::Clusters);
+
+using namespace o2::tpc::qc;
+
+//______________________________________________________________________________
+bool Clusters::processCluster(o2::tpc::ClusterNative const& cluster, o2::tpc::Sector const sector, const int row)
+{
+  const int nROC = row < 63 ? int(sector) : int(sector) + 36;
+  const int rocRow = row < 63 ? row : row - 63;
+
+  const float pad = cluster.getPad();
+
+  const uint16_t qMax = cluster.qMax;
+  const uint16_t qTot = cluster.qTot;
+  const float sigmaPad = cluster.getSigmaPad();
+  const float sigmaTime = cluster.getSigmaTime();
+  const float timeBin = cluster.getTime();
+
+  float count = mNClusters.getCalArray(nROC).getValue(rocRow, pad);
+  mNClusters.getCalArray(nROC).setValue(rocRow, pad, count + 1);
+
+  float charge = mQMax.getCalArray(nROC).getValue(rocRow, pad);
+  mQMax.getCalArray(nROC).setValue(rocRow, pad, charge + qMax);
+
+  charge = mQTot.getCalArray(nROC).getValue(rocRow, pad);
+  mQTot.getCalArray(nROC).setValue(rocRow, pad, charge + qTot);
+
+  count = mSigmaTime.getCalArray(nROC).getValue(rocRow, pad);
+  mSigmaTime.getCalArray(nROC).setValue(rocRow, pad, count + sigmaTime);
+
+  count = mSigmaPad.getCalArray(nROC).getValue(rocRow, pad);
+  mSigmaPad.getCalArray(nROC).setValue(rocRow, pad, count + sigmaPad);
+
+  count = mTimeBin.getCalArray(nROC).getValue(rocRow, pad);
+  mTimeBin.getCalArray(nROC).setValue(rocRow, pad, count + timeBin);
+
+  return true;
+}
+
+//______________________________________________________________________________
+void Clusters::analyse()
+{
+  mQMax /= mNClusters;
+  mQTot /= mNClusters;
+  mSigmaTime /= mNClusters;
+  mSigmaPad /= mNClusters;
+  mTimeBin /= mNClusters;
+}
+
+//______________________________________________________________________________
+void Clusters::dumpToFile(std::string filename)
+{
+  if (filename.find(".root") != std::string::npos) {
+    filename.resize(filename.size() - 5);
+  }
```

Comment:
If you used string_views they can easily be manipulated. Please check the documentation.

Reply:
see above

Reply:
Could work like this:
```c++
std::string_view filename = "test.root";
const auto rootpos = filename.rfind(".root");
if (rootpos != std::string_view::npos) {
  filename=filename.substr(0, rootpos);
}

std::string outFileBase(filename);

TFile::Open((outFileBase + "_canvas.root").c_str(), "recreate");
```
But as you like since this is totally unimportant...